Source/JSONAPISerialization.swift

```diff
@@ -0,0 +1,206 @@
+    
+    public func data(includeRelationships includeRelationships: Bool, includeAttributes: Bool) -> AnyObject? {
+        return Element.self is JSONAPISerializable.Type ? flatMap { ($0 as? JSONAPISerializable)?.data(includeRelationships: includeRelationships, includeAttributes: includeAttributes) } : nil
+    }
+}
+
+extension PropertyPolicy: JSONAPISerializable {
+
+    // MARK: JSONAPISerializable
+    
+    public func data(includeRelationships includeRelationships: Bool, includeAttributes: Bool) -> AnyObject? {
+        guard Wrapped.self is JSONAPISerializable.Type else {
+            return nil
+        }
+        
+        switch self {
+        case let .Some(value):
+            if let value = value as? JSONAPISerializable {
+                return value.data(includeRelationships: includeRelationships, includeAttributes: includeAttributes)
+            }
+            
+        case .Null:
+            return [String: AnyObject]() // included as relationship but empty
+        case .None:
+            return nil
+        }
+        
+        return nil
+    }
+}
+
+extension Optional: JSONAPISerializable {
+    
+    // MARK: JSONAPISerializable
+    
+    public func data(includeRelationships includeRelationships: Bool, includeAttributes: Bool) -> AnyObject? {
+        guard Wrapped.self is JSONAPISerializable.Type else {
+            return nil
+        }
+        
+        switch self {
+        case let .Some(value):
+            if let value = value as? JSONAPISerializable {
+                return value.data(includeRelationships: includeRelationships, includeAttributes: includeAttributes)
+            }
+        case .None:
+            return nil
+        }
+        
+        return nil
+    }
+}
+
+public extension JSONAPIEntity {
+    
+    // MARK: JSONAPIEntity
+
+    var type: String {
+        return String(self.dynamicType).lowercaseString
+    }
+    
+    // MARK: CustomSerializable
+
+    public func customSerialize() -> AnyObject? {
+        return data(includeRelationships: true, includeAttributes: true)!
+    }
+    
+    // MARK: JSONAPISerializable
+
+    public func data(includeRelationships includeRelationships: Bool, includeAttributes: Bool) -> AnyObject? {
+        var data = [String: AnyObject]()
+        
+        data["id"] = id
+        data["type"] = type
+        
+        guard includeRelationships || includeAttributes else {
+            return data
+        }
+        
+        let mirror = Mirror(reflecting: self)
+        
+        var attributes = [String: AnyObject]()
+        var relationships = [String: AnyObject]()
+        
+        for child in mirror.children {
+            if let label = child.label {
+                if let value = child.value as? JSONAPISerializable, let data = value.data(includeRelationships: false, includeAttributes: false) {
+                    if includeRelationships {
+                        relationships[label] =  ["data": data]
+                    }
```

Comment:
we can't avoid `value.data` because if it returns nil it means that value is an attribute (just added doc for this), we can avoid it only if `includeAttributes` is also nil but then we would never reach this point (when both boolean are nil).
For the second question I don't get it